mlos_bench/mlos_bench/optimizers/grid_search_optimizer.py

```diff
@@ -0,0 +1,151 @@
+        -------
+        List[HashableTunableValuesDict]
+        """
+        return self._pending_configs.keys()
+
+    @property
+    def suggested_configs(self) -> Set[HashableTunableValuesDict]:
+        """
+        Gets the set of configs that have been suggested but not yet registered.
+
+        Returns
+        -------
+        Set[HashableTunableValuesDict]
+        """
+        return self._suggested_configs
+
+    def bulk_register(self, configs: Sequence[dict], scores: Sequence[Optional[float]],
+                      status: Optional[Sequence[Status]] = None, is_warm_up: bool = False) -> bool:
+        if not super().bulk_register(configs, scores, status, is_warm_up):
+            return False
+        if status is None:
+            status = [Status.SUCCEEDED] * len(configs)
+        for (params, score, trial_status) in zip(configs, scores, status):
+            tunables = self._tunables.copy().assign(params)
+            self.register(tunables, trial_status, None if score is None else float(score))
+            if is_warm_up:
+                # Do not advance the iteration counter during warm-up.
+                self._iter -= 1
+        if _LOG.isEnabledFor(logging.DEBUG):
+            (score, _) = self.get_best_observation()
+            _LOG.debug("Warm-up end: %s = %s", self.target, score)
+        return True
+
+    def suggest(self) -> TunableGroups:
+        """
+        Generate the next grid search suggestion.
+        """
+        tunables = self._tunables.copy()
+        if self._start_with_defaults:
+            _LOG.info("Use default values for the first trial")
+            self._start_with_defaults = False
+            tunables = tunables.restore_defaults()
+            default_config = tunables.get_param_values()
+            # Move the default from the pending to the suggested set.
+            del self._pending_configs[default_config]
+            self._suggested_configs.add(default_config)
+        else:
+            # Select the first item from the pending configs.
+            next_config = next(iter(self._pending_configs.keys()))
+            tunables.assign(next_config)
+            # Move it to the suggested set.
+            self._suggested_configs.add(next_config)
+            del self._pending_configs[next_config]
+        _LOG.info("Iteration %d :: Suggest: %s", self._iter, tunables)
+        return tunables
+
+    def register(self, tunables: TunableGroups, status: Status,
+                 score: Optional[Union[float, dict]] = None) -> Optional[float]:
+        registered_score = super().register(tunables, status, score)
+        if status.is_succeeded() and (
+            self._best_score is None or (registered_score is not None and registered_score < self._best_score)
+        ):
+            self._best_score = registered_score
+            self._best_config = tunables.copy()
+        self._iter += 1
+        try:
+            self._suggested_configs.remove(tunables.get_param_values())
```

Comment:
How about we just keep track of the *tuples of values* instead of introducing a special data structure? I mean, something like
```suggestion
            self._suggested_configs.remove(tuple(tunables.get_param_values().values()))
```
(if you wish, we can also add some sanity check to make sure that the keys of the tunables match those we've passed to the optimizer in the constructor, but I think it is not necessary)

Reply:
Not a bad idea.  I'll poke at it.

Reply:
Comments on this in the other thread below.

Reply:
The short version is this is what I did, but you have to be careful about what the source of the dict is since the order of the values in the tuple matters.